code/modules/mechanics/MechanicsTextToMusicCom.dm

```diff
@@ -0,0 +1,478 @@
+#define MIN_TIMING 0.1
+#define MAX_TIMING 0.5
+
+#define MAX_NOTE_INPUT 1920
+#define MAX_CONCURRENT_NOTES 8
+
+#define CF_EVENT_LENGTH 8 // how long a Classic Format event is in characters
+
+#define DELAY_NOTE_MIN 0
+#define DELAY_NOTE_MAX 80
+#define DELAY_REST_MIN 1
+#define DELAY_REST_MAX 1000
+
+// Defines for the Dense Format (DF)
+#define DF_EVENT_LENGTH 3      // how long a DF event is in characters
+#define DF_DELIMITER 32        // space character (reserved char)
+#define DF_BOUND_LOWER 33      // !
+#define DF_BOUND_UPPER 120     // x
+#define DF_REST 121            // y
+#define DF_SET_TIMING 122      // z
+#define DF_OFFSET_DYNAMIC -13  // shift dynamic from 33 to 20
+#define DF_OFFSET_NOTE -12     // shift note    from 33 to 21
+#define DF_OFFSET_DELAY -33    // shift delay   from 33 to 0
+#define DF_OFFSET_REST -33     // shift rest    from 33 to 0
+#define DF_OFFSET_OCTAVE 1     //
+#define DF_NOTE_TYPE_AMOUNT 12 // the amount of note types there in an octave
+#define DF_DYNAMIC_MIN 20      // lowest  sound volume
+#define DF_DYNAMIC_MAX 60      // highest sound volume
+#define DF_FORBIDDEN_CHAR "|"  // no pipes please
+#define DF_BASE88 88           // how many characters there are in base88
+
+#define STANDARD_INSTRUMENT_PATH(instrument_name) "sound/musical_instruments/[instrument_name]/notes"
+#define MAX_ERROR_MESSAGES 5 // the most error messages allowed before older messages are deleted
+
+TYPEINFO(/obj/item/mechanics/text_to_music)
+	mats = list("metal"      = 20,
+				"conductive" = 10,
+				"crystal"    = 10)
+
+/obj/item/mechanics/text_to_music // modified from playable_piano.dm and PR #21051 (Player Piano Notes Rework V2)
```

Comment:
should be `cabinet_banned`. we really don't need these things walking around

Reply:
Sure